lib/linked_list.rb

```diff
@@ -7,39 +7,67 @@ def initialize
     @head = nil
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def add_first(data)
-
+    @head = Node.new(data, @head)
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(1)
+  # Space complexity - O(1)
   def get_first
-
+    return nil if @head == nil
+    return @head.data
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n), where n is the number of nodes in the linked list
+  # Space complexity - O(1)
   def length
-    return 0
+    return 0 if @head == nil
+    current = @head
+    length = 1
+    until current.next == nil
+      current = current.next
+      length += 1
+    end
+    return length
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n), where n is the number of nodes in the linked list
+  # Space complexity - O(1)
   def add_last(data)
-
+    if @head == nil
+      @head = Node.new(data, nil)
+    else
+      current = @head
+      until current.next == nil
+        current = current.next
+      end
+      current.next = Node.new(data, nil)
+    end
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - O(n), where n is the number of nodes in the linked list
+  # Space complexity - O(1)
   def get_last
-
+    current = @head
```

Comment:
What if `@head` is `nil`?

```suggestion
    return nil if @head.nil?
    current = @head
```